File: code/code3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ordSet = set()
        n = len(s)
        r = 0
        ans = 0
        for i in range(n):
            if i != 0:
                ordSet.remove(s[i-1])
            while r < n and s[r] not in ordSet:
                ordSet.add(s[r])
                r = r + 1
            ans = max(ans,r - i)
        return ans
if __name__ == "__main__":
    solution = Solution()
    s = "abcabcbb"
    ans = solution.lengthOfLongestSubstring(s)
    print(ans)

File: code/code3.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int right = 0;
        int len = s.size();
        unordered_set<char> uset;
        int ans = 0;
        for (int left = 0;left < len;left++) {
            if (left != 0) {
                uset.erase(s[left-1]);
            }
            while (right < len && !uset.count(s[right])) {
                uset.insert(s[right]);
                right++;
            }
            ans = max(ans,int(uset.size()));
        } 
        return ans;
    }
};
int main(){
    Solution solution;
    string s = "abcabcbb";
    int ans = solution.lengthOfLongestSubstring(s);
    cout << ans << endl;
}